### PR TITLE
Feat: refactor fixed-point math in `pox_5_make_reward_set`

### PR DESCRIPTION
**Pull request by aaronb-stacks**
This does a bit of refactoring of `pox_5_make_reward_set` and the `uint` module in `stacks-common`. In the first case, it introduces and uses a `FixedPointU256` struct to perform the reward set math -- this changed the precision, but the previous precision was pretty unnecessary (64 fractional bits is plenty).

The refactor of the wide uint types should bring some performance gains -- using rust's native multiplication, subtraction, and addition operations for supporting wider types (`carrying_*` and `borrowing_sub`) allowing for in-place modification for the self-consuming implementations, and using reference args for the cloning implementations. This avoids a fair number of clones in multiplication in particular.

* Add new FixedPointU256 struct to `uint` in common
* Use fewer bits of scaling -- 128 bits is totally unnecessary
* Add checked variants and some more efficient calculations to wide uints
* Add proptests and extreme value tests to ensure PoX calculations do not overflow

### CODE REVIEW

File: stacks-common/src/util/uint.rs [hunk 1]
```diff
@@ -189,86 +495,115 @@ macro_rules! construct_uint {
...
+        impl ::std::ops::Sub<&Self> for $name {
+            type Output = Self;
 
-            fn mul(self, other: $name) -> $name {
-                let mut me = $name::zero();
-                // TODO: be more efficient about this
-                for i in 0..(2 * $n_words) {
-                    let to_mul = (other >> (32 * i)).low_u32();
-                    me = me + (self.mul_u32(to_mul) << (32 * i));
-                }
```
Comment (aaronb-stacks) [inline stacks-common/src/util/uint.rs:260]: Each "add" here can do 2-4 clones of the wide type (because of how it implements carry and because, despite consuming the input arg, it allocates new arrays for the return and carry). Each shift is another clone, and mul_u32 also clones. This means that each iteration of the loop performs 5-7 clones, so multiplication ends up performing (10-14 * N_WORDS) clones.

**Review by brice-stacks — APPROVED** [review]
This looks great! 🙏

File: changelog.d/7087-fixed-point-struct.added
Comment (brice-stacks) [inline changelog.d/7087-fixed-point-struct.added:1]: Move this into changelog.d/ to make the check happy.

**Review by brice-stacks — APPROVED** [review]
👍

File: stacks-common/src/util/uint.rs [hunk 2]
```diff
@@ -48,92 +49,395 @@ pub trait BitArray {
...
+impl<const S: u16> std::fmt::Display for FixedPointU256<S> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{self:?}")
+    }
+}
+
+impl<const SCALE: u16> FixedPointU256<SCALE> {
+    /// Compile-time assertion that `SCALE < 128`, ensuring that even
+    /// `u128::MAX` can always be represented without overflow.
+    const _ASSERT_SCALE: () = assert!(SCALE < 128, "SCALE must be less than 128");
```
Comment (jcnelson) [inline stacks-common/src/util/uint.rs:137]: You might want to make this 64 instead of 128, since it impacts how big the numbers can be under multiplication.  For example, if `SCALE = 127`, then `FixedPointU256::from_u64(1) * FixedPointU256::from_u64(1)` would require computing `1 << 254`.  You couldn't even compute `FixedPointU256::from_u64(10) * FixedPointU256::from_u64(1)`.

File: stacks-common/src/util/uint.rs [hunk 3]
```diff
@@ -48,92 +49,395 @@ pub trait BitArray {
...
+        }
+        let value = self.value / other;
+        Some(Self { value })
+    }
+
+    /// Raise `self` to the power of `exp`, trimming the scale to `self.scaling` on each multiplication.
+    /// Returns None if overflowed
+    pub fn pow(&self, exp: u32) -> Option<Self> {
+        if exp == 0 {
+            return Some(Self::MINIMAL);
```
Comment (jcnelson) [inline stacks-common/src/util/uint.rs]: Shouldn't this be `Some(Self::from_u64(1))`?

File: stacks-common/src/util/uint.rs [hunk 4]
```diff
@@ -48,92 +49,395 @@ pub trait BitArray {
...
+        }
+        if n == 0 {
+            return None;
+        }
+
+        let mut low = Self::MINIMAL;
+        let mut high = self.clone();
+        loop {
+            if high <= low {
+                return Some(high.min(low));
```
Comment (jcnelson) [inline stacks-common/src/util/uint.rs:298]: Wouldn't this just be `high`?

File: stacks-common/src/util/uint.rs [hunk 5]
```diff
@@ -48,92 +49,395 @@ pub trait BitArray {
...
+            }
+            let guess = high.clone().add(&low)?.div(Uint256::from_u64(2))?;
+            match guess.pow(n) {
+                None => {
+                    // Overflow means guess is too large
+                    high = guess.sub(&Self::MINIMAL)?;
```
Comment (jcnelson) [inline stacks-common/src/util/uint.rs:304]: The fact that `Self::MINIMAL` is so small might mean that this loop here executes many, many times, right?

File: stacks-common/src/util/uint.rs [hunk 6]
```diff
@@ -48,92 +49,395 @@ pub trait BitArray {
...
+        let mut out = Vec::with_capacity(Uint256::N_WORDS);
+        for word in self.value.0.iter() {
+            out.extend((*word).to_le_bytes());
+        }
+        out
+    }
+
+    /// Find the `n`th root of self using binary search, trimming the scale to `self.scaling` on
+    ///  iterations
+    pub fn find_root_floor(&self, n: u32) -> Option<Self> {
```
Comment (jcnelson) [inline stacks-common/src/util/uint.rs:286]: I get that the algorithm calls for computing a weighted geomean, but you don't need to employ an nth-root function if you have an integer log-2 function (which I think is easier to come by and less risky to implement).  I mentioned this elsewhere, but I'll repost it here:

```
You don't have to take the 15th-root of anything; you could instead compute this:

                      (  5 * ilog2(STX0) + 4 * ilog2(STX1) + 3 * ilog2(STX2) + 2 * ilog2(STX3) + ilog2(STX4) )
                      ( ------------------------------------------------------------------------------------ )
                      (  5 * ilog2(BTC0) + 4 * ilog2(BTC1) + 3 * ilog2(BTC2) + 2 * ilog2(BTC3) + ilog2(BTC4) )
 GeoMean(STX/BTC) = 2

where ilog2(x) is the integer base-2 logarithm of x (which is easy to compute on fix-point numbers).
```

The nice thing IMHO about `ilog2(x)` is that because its range is bounded for all reasonable values of `x`, it's trivial to implement as a look-up table.  If needed, you could interpolate between two points in the table for extra precision.

File: stacks-common/src/util/uint.rs (hunk 2 above)
Comment (aaronb-stacks) [reply stacks-common/src/util/uint.rs:137]: I think that's fine, though. Checked math would return `None` in those cases, and the other options would wrap. This assertion is basically here to avoid needing to make `from_u128()` and `from_u64()` fallible.

File: stacks-common/src/util/uint.rs (hunk 6 above)
Comment (aaronb-stacks) [reply stacks-common/src/util/uint.rs:286]: Yeah, that's better.

File: stacks-common/src/util/uint.rs (hunk 3 above)
Comment (aaronb-stacks) [reply stacks-common/src/util/uint.rs]: Yes

File: stacks-common/src/util/uint.rs (hunk 6 above)
Comment (aaronb-stacks) [reply stacks-common/src/util/uint.rs:286]: Hmm. This means dealing with negative values, right?

File: stacks-common/src/util/uint.rs (hunk 5 above)
Comment (aaronb-stacks) [reply stacks-common/src/util/uint.rs:304]: I don't think this impacts how many times the loop executes -- it still halves the `low, high` range on every iteration.

File: stacks-common/src/util/uint.rs (hunk 6 above)
Comment (jcnelson) [reply stacks-common/src/util/uint.rs:286]: If the units of STX and BTC are uSTX and satoshis, respectively, then the range of `ilog2(x)` would always be positive since you can't have a fractional uSTX or satoshi.

File: stacks-common/src/util/uint.rs [hunk 7]
```diff
@@ -48,92 +49,395 @@ pub trait BitArray {
...
+            return None;
+        }
+
+        let mut low = Self::MINIMAL;
+        let mut high = self.clone();
+        loop {
+            if high <= low {
+                return Some(high.min(low));
+            }
+            let guess = high.clone().add(&low)?.div(Uint256::from_u64(2))?;
```
Comment (jcnelson) [inline stacks-common/src/util/uint.rs:300]: You could perhaps speed up this algorithm if your initial guess was simply computed from an `f64`, cast to a fixed-point.

File: stacks-common/src/util/uint.rs (hunk 5 above)
Comment (jcnelson) [reply stacks-common/src/util/uint.rs:304]: Ah, yes

File: stacks-common/src/util/uint.rs (hunk 6 above)
Comment (jcnelson) [reply stacks-common/src/util/uint.rs:286]: Actually, I'm not sure how much this alternative representation saves us in complexity. You're still computing a fractional power of 2 in essentially all cases, which necessitates an nth-root fixed point function (unless there's a fast fixed-point algorithm that works specifically for fractional powers of 2).

Comment (aaronb-stacks) [reply stacks-common/src/util/uint.rs:286]: > If the units of STX and BTC are uSTX and satoshis, respectively, then the range of ilog2(x) would always be positive since you can't have a fractional uSTX or satoshi.

Oh, right -- I was figuring that I'd still operate on the ratios, but yeah, if I use the raw totals, I can avoid that.

> Actually, I'm not sure how much this alternative representation saves us in complexity. You're still computing a fractional power of 2 in essentially all cases, which necessitates an nth-root fixed point function (unless there's a fast fixed-point algorithm that works specifically for fractional powers of 2).

Yes there are fast fixed-point algorithms for fractional powers of 2, basically successive square roots:

For a fractional binary number `f = 0.101011` (most significant bit first)
```
2^f = 2^(1/2)*2^(1/8)*2^(1/32)*2^(1/64)
```
i.e., `2^f` is equal to the product of the n-th successive square root of two for each bit set in f. Because we're just using successive square roots of 2, those can be stored in a 64-entry table, so the algorithm basically becomes a bit shift to find the exp2 of the integer part, then (at-most) 64 multiplications to find the exp2 of the fractional part.

This is definitely more efficient, but I feel like it ends up being more complex -- the root finding algorithm may be slow (though I agree it could be sped up with the floating-point guess as you mentioned), but binary search seems comparatively simple.

File: stacks-common/src/util/uint.rs (hunk 4 above)
Comment (aaronb-stacks) [reply stacks-common/src/util/uint.rs:298]: That's equivalent here, but in my mind, the stopping condition for binary search is always "if high and low are equal or have crossed, return the minimum of the bounds."

File: stacks-common/src/util/uint.rs (hunk 6 above)
Comment (aaronb-stacks) [reply stacks-common/src/util/uint.rs:286]: I've implemented and benchmarked the two different approaches -- I'm going to open a separate PR for optimizations, benchmarks, and stress tests, because I don't want to stall this PR with discussion.

File: stacks-common/src/util/uint.rs (hunk 7 above)
Comment (aaronb-stacks) [reply stacks-common/src/util/uint.rs:300]: I have an implementation that does this, but will open in a separate PR with benchmarks, etc.
